lib/active_merchant/billing/gateways/reach.rb

```diff
@@ -110,11 +110,41 @@ def add_customer_data(request, options, payment)
         }.compact
       end
 
-      def add_custom_fields_data(request, options)
-        if options[:device_fingerprint].present?
-          request[:DeviceFingerprint] = options[:device_fingerprint]
-          request[:ViaAgent] = false
+      def add_stored_credentials(request, options)
+        request[:PaymentModel] = add_payment_model(options)
+        if request[:PaymentModel].nil?
+          raise ArgumentError,
+            "stored credential options doesn't match with PaymentModels combination list, you can also pass in the options directly the PaymentModels value"
         end
+
+        if request[:PaymentModel].match?(/CIT-/)
+          request[:DeviceFingerprint] = options[:device_fingerprint] if options[:device_fingerprint]
+        end
+      end
+
+      def add_payment_model(options)
+        return options[:payment_model] if options[:payment_model]
```

Comment:
💬  Comment:

Is a pretty particualr implementation and works, going to leave you a hash based alternative for your consideration => https://gist.github.com/54e9832c74bce33de4e84a49b8b85e29